src/main/java/net/dv8tion/jda/core/JDABuilder.java

```diff
@@ -539,7 +558,10 @@ public JDA buildAsync() throws LoginException, IllegalArgumentException, RateLim
     {
         OkHttpClient.Builder httpClientBuilder = this.httpClientBuilder == null ? new OkHttpClient.Builder() : this.httpClientBuilder;
         WebSocketFactory wsFactory = this.wsFactory == null ? new WebSocketFactory() : this.wsFactory;
-        JDAImpl jda = new JDAImpl(accountType, httpClientBuilder, wsFactory, shardRateLimiter, autoReconnect, enableVoice, enableShutdownHook,
+
+        if (controller == null && (reconnectQueue != null || shardRateLimiter != null))
+            controller = new ProvidingSessionController();
```

Comment:
Unsure if this is worth it, might be better to completely drop the legacy system.